clients/ui/bff/internal/integrations/k8s.go

```diff
@@ -26,62 +32,157 @@ type ServiceDetails struct {
+			BindAddress: "0", // disable metrics serving
+		},
+		HealthProbeBindAddress: "0", // disable health probe serving
+		LeaderElection:         false,
+		//Namespace:              "namespace", //TODO (ederign) do we need to specify the namespace to operate in
+		//There is also cache filters and Sync periods to assess later.
+	})
+
 	if err != nil {
-		return nil, fmt.Errorf("failed to create Kubernetes clientset: %w", err)
+		logger.Error("unable to create manager", "error", err)
+		cancel()
+		os.Exit(1)
 	}
-	//fetching services
-	services, err := clientSet.CoreV1().Services(namespace).List(context.TODO(), metav1.ListOptions{})
-	if err != nil {
-		return nil, fmt.Errorf("failed to list model-registry-server services: %w", err)
+
+	// Channel to signal when the manager has stopped
+	mgrStopped := make(chan struct{})
+
+	// Start the manager in a goroutine
+	go func() {
+		defer close(mgrStopped) // Signal that the manager has stopped
+		if err := mgr.Start(ctx); err != nil {
+			logger.Error("problem running manager", "error", err)
+			os.Exit(1)
+		}
+	}()
+
+	// Wait for the cache to sync before using the client
+	if !mgr.GetCache().WaitForCacheSync(ctx) {
+		cancel()
+		return nil, fmt.Errorf("failed to wait for cache to sync")
 	}
 
-	//building serviceCache
-	serviceCache, err := buildModelRegistryServiceCache(logger, *services)
+	kc := &KubernetesClient{
+		Client:     mgr.GetClient(),
+		Mgr:        mgr,
+		Token:      kubeconfig.BearerToken,
+		Logger:     logger,
+		stopFn:     cancel,
+		mgrStopped: mgrStopped, // Store the stop channel
+
+		//Namespace:    namespace, //TODO (ederign) do we need to restrict service list by namespace?
+	}
+	return kc, nil
+}
+
+func (kc *KubernetesClient) Shutdown(ctx context.Context, logger *slog.Logger) error {
+	logger.Info("shutting down Kubernetes manager...")
+
+	// Use the saved cancel function to stop the manager
+	kc.stopFn()
+
+	// Wait for the manager to stop or for the context to be canceled
+	select {
+	case <-kc.mgrStopped:
+		logger.Info("Kubernetes manager stopped successfully")
+		return nil
+	case <-ctx.Done():
+		logger.Error("context canceled while waiting for Kubernetes manager to stop")
+		return ctx.Err()
+	case <-time.After(30 * time.Second):
+		logger.Error("timeout while waiting for Kubernetes manager to stop")
+		return fmt.Errorf("timeout while waiting for Kubernetes manager to stop")
+	}
+}
+
+func (kc *KubernetesClient) BearerToken() (string, error) {
+	return kc.Token, nil
+}
+
+func (kc *KubernetesClient) GetServiceNames() ([]string, error) {
+	//TODO (ederign) when we develop the front-end, implement subject access review here
+	// and check if the username has actually permissions to access that server
+	// currently on kf dashboard, the user name comes in kubeflow-userid
+
+	//TODO (ederign) we should consider and rethinking listing all services on cluster
+	// what if we have thousand of those?
+	// we should consider label filtering for instance
```

Comment:
So does the code below return every service on the cluster and then our loop pulls out anything with component == "model-registry-server"?

Reply:
For now temporarly yes. I just started a slack thread with everyone to discuss the implications of this.